Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs

```diff
@@ -0,0 +1,304 @@
+    }
+
+    private void UpdateUiState(EntityUid uid, EntityUid loaderUid, MessagesCartridgeComponent? component)
+    {
+        if (!Resolve(uid, ref component))
+            return;
+        if (!TryComp(uid, out CartridgeComponent? cartComponent))
+            return;
+        MessagesUiState state;
+        MapId mapId = Transform(uid).MapID;
+        int? currentUserId = GetUserUid(cartComponent);
+        if (currentUserId == null || component.LastServer == null)
+        {
+            state = new MessagesUiState(MessagesUiStateMode.Error, [], null);
+            _cartridgeLoaderSystem.UpdateCartridgeUiState(loaderUid, state);
+            return;
+        }
+        if (component.ChatUid == null) //if no chat is loaded, list users
+        {
+            List<(string, int?)> userList = [];
+
+            var nameDict = _messagesServerSystem.GetNameDict(component.LastServer);
+
+            foreach (var nameEntry in nameDict.Keys)
+            {
+                if (nameEntry == currentUserId)
+                    continue;
+                userList.Add((nameDict[nameEntry], nameEntry));
+            }
+
+            userList.Sort
+            (
+                delegate ((string, int?) a, (string, int?) b)
+                {
+                    return String.Compare(a.Item1, b.Item1);
+                }
+            );
+
+            state = new MessagesUiState(MessagesUiStateMode.UserList, userList, null);
+        }
+        else
+        {
+            List<MessagesMessageData> messageList = []; //Else, list messages from the current chat
+
+            foreach (var message in _messagesServerSystem.GetMessages(component.LastServer, component.ChatUid.Value, currentUserId.Value))
+            {
+                if (message.SenderId == component.ChatUid && message.ReceiverId == currentUserId || message.ReceiverId == component.ChatUid && message.SenderId == currentUserId)
+                {
+                    messageList.Add(message);
+                }
+            }
+
+            messageList.Sort
+            (
+                delegate (MessagesMessageData a, MessagesMessageData b)
+                {
+                    return TimeSpan.Compare(a.Time, b.Time);
+                }
+            );
+
+            List<(string, int?)> formattedMessageList = [];
+
+            foreach (var message in messageList)
+            {
+                var name = GetName(message.SenderId, component);
+                var stationTime = message.Time.Subtract(_gameTicker.RoundStartTimeSpan);
+                var content = $"{stationTime.ToString("\\[hh\\:mm\\:ss\\]")} {name}: {message.Content}";
+                formattedMessageList.Add((content, null));
+            }
+
+            state = new MessagesUiState(MessagesUiStateMode.Chat, formattedMessageList, GetName(component.ChatUid.Value, component));
+        }
+        _cartridgeLoaderSystem.UpdateCartridgeUiState(loaderUid, state);
+    }
+
+}
```

Comment:
Why do we need to send entire UI states if the user changes who they're chatting too? Their message history isn't going to be changing so this is just going to waste data, especially if they decide to view it again.

You're better off just using component states + SessionSpecific instead of a BUI state, then dirtying the cartridge if it gets loaded or not. Then whenever the messages change you can just dirty the component instead of trying to add handlers around this like you did above.